src/util/thread_pool.h

```diff
@@ -0,0 +1,210 @@
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#ifndef UTIL_THREAD_POOL_HEADER
+#define UTIL_THREAD_POOL_HEADER
+
+#include "../threading/mutex.h"
+#include "../threading/mutex_auto_lock.h"
+#include "../threading/semaphore.h"
+#include "../threading/thread.h"
+
+class PoolableUpdateThread : public Thread
+{
+public:
+	PoolableUpdateThread(const std::string &name, Semaphore *sem,
+		Mutex *quitcount_mutex, u32 *quitcount) :
+		Thread(name + "Update"),
+		m_sem(sem),
+		m_quitcount_mutex(quitcount_mutex),
+		m_quitcount(quitcount) {}
+	~PoolableUpdateThread() {}
+
+	void *run()
+	{
+		DSTACK(FUNCTION_NAME);
+		BEGIN_DEBUG_EXCEPTION_HANDLER
+
+		while (!stopRequested()) {
+			m_sem->wait();
+
+			if (stopRequested())
+				break;
+
+			{
+				MutexAutoLock lock(*m_quitcount_mutex);
+				if (m_quitcount < 0) {
+					--m_quitcount;
+					break;
+				}
+			}
+
+			doUpdate();
+		}
+
+		END_DEBUG_EXCEPTION_HANDLER
+
+		return NULL;
+	}
+
+protected:
+	virtual void doUpdate() = 0;
+
+private:
+	Semaphore *m_sem;
+	Mutex *m_quitcount_mutex;
+	u32 *m_quitcount;
+};
+
+class ThreadPool
+{
+public:
+	ThreadPool(const std::string &name, size_t thread_count = 1) :
+		m_name(name),
+		m_threads(thread_count < 0 ? 0 : thread_count),
+		m_quitcount(0)
+	{
+		if (thread_count < 0)
+			thread_count = 0;
+
+		for (size_t i = 0; i < thread_count; ++i) {
+			spawnThread();
+		}
+	}
+	~ThreadPool()
+	{
+		stopAllThreadsWait();
+	}
+
+	void stopAllThreads()
+	{
+		for (size_t i = 0; i < m_threads.size(); ++i) {
+			if (m_threads[i]->isRunning()) {
+				m_threads[i]->stop();
+				stopThread();
+			}
+		}
+	}
+
+	void stopAllThreadsWait()
+	{
+		stopAllThreads();
+		size_t i = runningThreadCount();
+		while (i--) {
+			m_threads[i]->wait();
+		}
+	}
+
+	void setRunningThreadCount(size_t count)
+	{
+		if (count < 0)
+			count = 0;
+
+		const size_t thread_count = runningThreadCount();
+		if (thread_count == count)
+			return;
+
+		if (thread_count < count) {
+			for (size_t i = thread_count; i < count; ++i)
+				spawnThread();
+		} else /* (m_thread_count > count) */ {
+			for (size_t i = count; i < thread_count; ++i)
+				stopThread();
+		}
+	}
+
+	size_t threadCount() const
+	{
+		return m_threads.size();
+	}
+
+	size_t runningThreadCount() const
+	{
+		size_t running = 0;
+		for (size_t i = 0; i < m_threads.size(); ++i) {
+			if (m_threads[i]->isRunning())
+				++running;
+		}
+		return running;
+	}
+
+	inline bool isRunning() const
+	{
+		return runningThreadCount() > 0;
+	}
+
+	PoolableUpdateThread &getThread(size_t id)
+	{
+		return *m_threads[id];
+	}
+
+	void nudge()
+	{
+		m_sem.post();
+	}
+
+protected:
+	virtual PoolableUpdateThread *makeThread() = 0;
+
+	int findFreeIndex(bool invert = false)
```

Comment:
The type is wrong and should've been a **signed** counterpart to `size_t`, since this method returns `-1` when no thread slot is found.